Selection of memory access methods, aampostincrement detection

sobuch opened the pull request:
New configuration command introduced: "riscv set_mem_access". It allows to specify which RISC-V memory access methods (progbuf, sysbus and/or abstract access) should be tried and in which order of priority.

Command "riscv set_prefer_sba" is left and works in backward compatible way, but is marked as deprecated.

Support for the optional aampostincrement bit is detected the first time abstract memory access is used. This enables the use of OpenOCD with HW without support for aampostincrement.

sobuch commented:
Unsupported postincrement was tested on SweRV EH1 1.7, example of unsupported but "buggy" postincrement (address is not incremented but abstract command doesn't fail) is SweRV EH1 1.6. Help with testing with fully supported postincrement would be appreciated.

sobuch commented:
What is the expected behavior when multiple memory access methods are available and first one fails? Currently after first failure requested read/write fails as well. However, here https://github.com/riscv/riscv-openocd/issues/445#issuecomment-582595208 seems to be described a different behavior - that after a failure another method is tried, if possible. We implemented the second version, as it seems to be an improvement. Would this change of behavior be acceptable?

JanMatCodasip commented:
To further add on & rephrase what @sobuch already mentioned, this pull request deals with two items:

* Selection of memory access methods - merely an extension of current command `riscv set_prefer_sba`. It is motivated by SweRV cores, for which it is needed to have the option to use abstract memory access with priority over system bus (and that is not possible in current OpenOCD).

* Detection of whether `aampostincrement` bit is supported by the target - already discussed in https://github.com/riscv/riscv-debug-spec/issues/531

Any comments are appreciated. Most importantly, the answer to @sobuch's question above - how to treat multiple memory access methods?

* If a memory access method is enabled and supported (e.g. access via program buffer & the progbuf is large enough), but the access fails, what is the preferred course of action?
  * Should we try the next memory access method on the list? (this is what this branch does, and what is mentioned in #445)
  * Or should we give up and report an error? (this is what the current OpenOCD code does)

Thank you.

JanMatCodasip commented:
To further add on & rephrase what @sobuch already mentioned, this pull request deals with two items:

* Selection of memory access methods - merely an extension of current command `riscv set_prefer_sba`. It is motivated by SweRV cores, for which it is needed to have the option to use abstract memory access with priority over system bus (and that is not possible in current OpenOCD).

* Detection of whether `aampostincrement` bit is supported by the target - already discussed in https://github.com/riscv/riscv-debug-spec/issues/531

Any comments are appreciated. Most importantly, the answer to @sobuch's question above - how to treat multiple memory access methods?

* If a memory access method is enabled and supported (e.g. access via program buffer & the progbuf is large enough), but the access fails, what is the preferred course of action?
  * Should we try the next memory access method on the list? (this is what this branch does, and what is mentioned in #445)
  * Or should we give up and report an error? (this is what the current OpenOCD code does)

Thank you.

timsifive commented:
There are different reasons an access might fail:
1. The access may not be supported depending on the current target state. E.g. progbuf accesses won't work while the target is running. In that case we'd want to fall back to the next option.
2. The access might not work for that particular region of memory. E.g. an abstract access might fail because the current hart doesn't have permission to access the memory, but an sba access would succeed. Falling back would be helpful there also.
3. There may be some other problem that indicates a real issue, and in that case we should probably abort.

To handle the first problem, I'd introduce a new ERROR_ code (if necessary) that can be returned to indicate that this access mode won't work right now (but might at a later time).

To handle the second, the real solution is to let the user select the access method for different address regions. That's getting more complex, and I'm not sure if it would see much use. The downside of not doing that is now every time there's an access to a certain reason we first try a method that won't work before falling back on one that does. This PR is already large enough. Even if you do want to implement that, let's do it in a separate PR.

Differentiating between the 2nd and 3rd might also be tricky, so let's handle them together as follows:
When an access fails, log a warning and fall back to the next method. If all enabled methods fail, log an error and fail. To prevent tons of warnings in case 2 above, track a flag for each access method to only print the warning once per access method. The flags are reset when the new `set_mem_access` command is used.

How does that sound?

sobuch commented:
We have something (kind of) similar to the real solution for second problem implemented in https://github.com/Codasip/swerv-openocd, but I don't think we currently have the resources to do that properly.

Could you please explain what's the added value of new error code? it seems to me that if we won't differentiate between the last two kinds now, we may just as well process the first one the same way.

Some flags to prevent warnings definitely seem like a good idea.

JanMatCodasip commented:
@timsifive - Tim, thank you for your reply.

As for handling of memory access failures - 

> To handle the first problem, I'd introduce a new ERROR_ code (if necessary) that can be returned to indicate that this access mode won't work right now (but might at a later time).

For progbuf accesses, this is easy - we can skip this access method if the target is running. We'll add this to the merge request.

For abstract memory access, this is more complex - it may or may not be supported while the target runs. In order to detect what the target supports, we would have to try the abstract command, and then possibly remember the outcome. This would be similar logic to what we've implemented here for the "aampostincrement" bit. For simplicity & robustness, I recommend not to add further intelligence that would try to remember what abstract commands are supported under what circumstances, at least not for now. This can be further enhanced in the future, if needed. 

So our proposal for abstract memory commands is: simply try them every time, as it is in the current trunk code. 

> Differentiating between the 2nd and 3rd might also be tricky, so let's handle them together as follows: When an access fails, log a warning and fall back to the next method. If all enabled methods fail, log an error and fail.  (....)

This sounds very fine. We will adjust the code this way.

> To handle the second, the real solution is to let the user select the access method for different address regions. That's getting more complex, and I'm not sure if it would see much use.

We're also not sure about the value of different memory regions in upstream OpenOCD. Either way, we should be fine even if the need arises at some point, because the current implementation can be extended in a backward-compabible manner - e.g. command `set_mem_access_region` (or something along those lines) would be introduced to specify custom methods for a particular range of addresses. `set_mem_access` would remain in place to specify the defaults - the methods to use outside of any defined region.


We will adjust the code within the next week and will reach back when done.

Thanks and regards,
Jan

timsifive commented:
> I recommend not to add further intelligence that would try to remember what abstract commands are supported under what circumstances, at least not for now. This can be further enhanced in the future, if needed.

That sounds good.

Thanks for the update.

sobuch committed:
Add flexible selection of memory access methods, detection of aampostincrement.

New configuration command introduced: "riscv set_mem_access".
It allows to specify which RISC-V memory access methods (progbuf,
sysbus and/or abstract access) should be tried and in which order
of priority.

Command "riscv set_prefer_sba" is left and works in backward
compatible way, but is marked as deprecated.

First time abstract memory access is executed, it is tried with
set aampostincrement bit. If the abstract command fails or the
address is not incremented correctly, aampostincrement will not
be used for any subsequent accesses.

Signed-off-by: Samuel Obuch <sobuch@codasip.com>

sobuch commented:
Hi @timsifive, we finished the implementation as specified here earlier. Riscv debug tests are passing, manual testing now also doesn't show any further problems. We look forward to any feedback from you, please let us known if there is anything more we should do before this pull request can be closed. Otherwise, this branch is rebased and changes squashed, so from our side it's ready to be merged :)

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -98,6 +98,9 @@ static int	riscv013_test_compliance(struct target *target);
 
 #define RISCV013_INFO(r) riscv013_info_t *r = get_info(target)
 
+#define LOG_WARNING_OR_DEBUG(use_warn, expr ...) \
```
You only use this macro twice, and in both places you have almost the same logic to generate the string.

Can you make this a regular function, which is only used for memory accesses, and takes arguments for read/write (bool read), bool success, and method?

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -1831,27 +1836,41 @@ static int riscv013_hart_count(struct target *target)
+			/* No further mem access method to try. */
+			break;
+	}
+	LOG_DEBUG("Unable to determine supported data bits on this target. Assuming 32 bits.");
```
Should this be an error? I don't see how we get here on a normal system.

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -2708,16 +2721,39 @@ static int read_memory_abstract(struct target *target, target_addr_t address,
+					LOG_DEBUG("aampostincrement is supported on this target.");
+					info->has_aampostincrement = YNM_YES;
+				} else {
+					LOG_DEBUG("Buggy aampostincrement! Address not incremented correctly.");
```
A warning would be more appropriate. Especially since the user will only see it once per session (right?).

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -2708,16 +2721,39 @@ static int read_memory_abstract(struct target *target, target_addr_t address,
+			}
+		}
+
 		if (result != ERROR_OK) {
```
Remove the braces, per OpenOCD style guide. I'm surprised checkpatch didn't fail the build for this.

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -3219,30 +3268,137 @@ static int read_memory(struct target *target, target_addr_t address,
+		int method = r->mem_access_methods[i];
+
+		if (method == RISCV_MEM_ACCESS_PROGBUF) {
+			if (!has_sufficient_progbuf(target, 3)) {
+				LOG_DEBUG("Skipping mem read via progbuf - insufficient progbuf size.");
+				progbuf_result = "skipped";
+				continue;
+			}
+			if (target->state != TARGET_HALTED) {
+				LOG_DEBUG("Skipping mem read via progbuf - target not halted.");
+				progbuf_result = "skipped";
+				continue;
+			}
+			if (riscv_xlen(target) < size * 8) {
+				LOG_DEBUG("Skipping mem read via progbuf - XLEN (%d) is too short for %d-bit memory read.",
+						riscv_xlen(target), size * 8);
+				progbuf_result = "skipped";
+				continue;
+			}
+			if (size > 8) {
+				LOG_DEBUG("Skipping mem read via progbuf - unsupported size.");
+				progbuf_result = "skipped";
+				continue;
+			}
+			if ((sizeof(address) * 8 > riscv_xlen(target)) && (address >> riscv_xlen(target))) {
+				LOG_DEBUG("Skipping mem write via progbuf - progbuf only supports %u-bit address.",
+						riscv_xlen(target));
+				progbuf_result = "skipped";
+				continue;
+			}
```
This looks identical to the code in write_memory_progbuf (except I suppose the read/write strings). Can you refactor this?

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -3219,30 +3268,137 @@ static int read_memory(struct target *target, target_addr_t address,
+				r->mem_access_progbuf_warn = false;
+			}
+		} else if (method == RISCV_MEM_ACCESS_SYSBUS) {
+			if ((!get_field(info->sbcs, DM_SBCS_SBACCESS8) || size != 1) &&
+					(!get_field(info->sbcs, DM_SBCS_SBACCESS16) || size != 2) &&
+					(!get_field(info->sbcs, DM_SBCS_SBACCESS32) || size != 4) &&
+					(!get_field(info->sbcs, DM_SBCS_SBACCESS64) || size != 8) &&
+					(!get_field(info->sbcs, DM_SBCS_SBACCESS128) || size != 16)) {
+				LOG_DEBUG("Skipping mem read via system bus - unsupported access size.");
+				sysbus_result = "skipped";
+				continue;
+			}
+			unsigned sbasize = get_field(info->sbcs, DM_SBCS_SBASIZE);
+			if ((sizeof(address) * 8 > sbasize) && (address >> sbasize)) {
+				LOG_DEBUG("Skipping mem read via system bus - sba only supports %u-bit address.", sbasize);
+				sysbus_result = "skipped";
+				continue;
+			}
+			if (increment != size && (get_field(info->sbcs, DM_SBCS_SBVERSION) == 0 || increment != 0)) {
+				LOG_DEBUG("Skipping mem read via system bus - "
+						"sba reads only support size==increment or also size==0 for sba v1.");
+				sysbus_result = "skipped";
+				continue;
+			}
```
Same.

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -3219,30 +3268,137 @@ static int read_memory(struct target *target, target_addr_t address,
+				r->mem_access_sysbus_warn = false;
+			}
+		} else if (method == RISCV_MEM_ACCESS_ABSTRACT) {
+			if (size > 8) {
+				/* TODO: Add 128b support if it's ever used. Involves modifying
+						 read/write_abstract_arg() to work on two 64b values. */
+				LOG_DEBUG("Skipping mem read via abstract access - unsupported size: %d bits", size * 8);
+				abstract_result = "skipped";
+				continue;
+			}
+			if ((sizeof(address) * 8 > riscv_xlen(target)) && (address >> riscv_xlen(target))) {
+				LOG_DEBUG("Skipping read write via abstract access - abstract access only supports %u-bit address.",
+						riscv_xlen(target));
+				abstract_result = "skipped";
+				continue;
+			}
+			if (size != increment) {
+				LOG_ERROR("Skipping mem read via abstract access - "
+						"abstract command reads only support size==increment.");
+				abstract_result = "skipped";
+				continue;
+			}
```
Same again.

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -3219,30 +3268,137 @@ static int read_memory(struct target *target, target_addr_t address,
+		if (method == RISCV_MEM_ACCESS_PROGBUF) {
+			if (!has_sufficient_progbuf(target, 3)) {
+				LOG_DEBUG("Skipping mem read via progbuf - insufficient progbuf size.");
+				progbuf_result = "skipped";
```
It looks like you're just tracking this string for the error messages. Can you give unique reasons instead of "skipped" for each one? E.g. "skipped (insufficient progbuf size)" etc.

timsifive commented on src/target/riscv/riscv.c:
```diff
@@ -2287,11 +2286,85 @@ COMMAND_HANDLER(riscv_test_compliance) {
+	struct target *target = get_current_target(CMD_CTX);
+	RISCV_INFO(r);
+	bool prefer_sba;
+	LOG_WARNING("`riscv set_prefer_sba` is deprecated. Please use `riscv set_mem_access` instead.");
```
Thanks for making this backwards compatible.

timsifive commented on src/target/riscv/riscv.c:
```diff
@@ -2287,11 +2286,85 @@ COMMAND_HANDLER(riscv_test_compliance) {
+{
+	struct target *target = get_current_target(CMD_CTX);
+	RISCV_INFO(r);
+	unsigned int i;
```
You're only using `i` inside loops. Can you declare it in each for loop instead? E.g.:
	for (unsigned int i = 0; i < CMD_ARGC; i++) {

timsifive reviewed (changes requested):
This structure looks great overall. I have some stylistic comments, but I like this design.

Also, please document the new command alongside the other riscv commands in https://github.com/riscv/riscv-openocd/blob/riscv/doc/openocd.texi

sobuch commented:
Thank you for looking into this and for the feedback. Looking at some of the read/write functions, probably more extensive refactor is in order. I tried at least not make it worse with this change, your suggestions were very helpful in this.

Otherwise, same as before - rebased, debug tests passing, checkpatch also passing

timsifive commented on src/target/riscv/riscv-013.c:
```diff
@@ -3234,30 +3413,65 @@ static int read_memory(struct target *target, target_addr_t address,
+		int method = r->mem_access_methods[i];
+
+		if (method == RISCV_MEM_ACCESS_PROGBUF) {
+			bool skip = mem_should_skip_progbuf(target, address, size, true, &progbuf_result);
+			if (skip)
+				continue;
```
The skip variable isn't used outside of here, right?

I'd just write:
```
if (mem_should_skip...)
   continue;
```

sobuch replied:
of course, I missed this

timsifive reviewed (dismissed):
This looks good. Thank you. I just have one last request below.

Also, instead of rebase+force push, can you `merge riscv` and regular push? That way when I'm reviewing code I can easily see just the changes you made since the last time I read it.

sobuch committed:
remove unnecessary variable

sobuch commented:
Thanks for the review. Next time I won't rebase, if you prefer it that way. Now I just added a new commit.

sobuch committed:
fix doc

timsifive reviewed (approved):
I'm happy with this. Thank you. I'll merge it tomorrow unless there are more comments.
